lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -674,83 +892,444 @@ export default function getAPIImplementation(): APIInterface {
+                    if (createPrimaryRes.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+                        // Session user is linked to another primary user, which means the session is revoked as well
 
-            // this function also does account linking
-            let response = await options.recipeImplementation.signUp({
-                tenantId,
-                email,
-                password,
-                userContext,
-            });
-            if (response.status === "EMAIL_ALREADY_EXISTS_ERROR") {
-                return response;
-            }
-            let emailPasswordRecipeUser = response.user.loginMethods.find(
-                (u) => u.recipeId === "emailpassword" && u.hasSameEmailAs(email)
-            );
+                        // We cannot recurse here because when the session user if fetched again,
+                        // it will be a primary user and we will end up trying factor setup with that user
+                        // Also this session would have been revoked and we won't be able to catch it again
 
-            if (emailPasswordRecipeUser === undefined) {
-                // this can happen cause of some race condition, but it's not a big deal.
-                throw new Error("Race condition error - please call this API again");
-            }
+                        throw new SessionError({
+                            type: SessionError.UNAUTHORISED,
+                            message: "Session may be revoked",
+                        });
+                    } else if (
+                        createPrimaryRes.status === "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR"
+                    ) {
+                        throw new RecurseError();
+                    }
+                }
 
-            let session = await Session.createNewSession(
-                options.req,
-                options.res,
-                tenantId,
-                emailPasswordRecipeUser.recipeUserId,
-                {},
-                {},
-                userContext
-            );
-            return {
-                status: "OK",
-                session,
-                user: response.user,
+                const linkRes = await AccountLinking.getInstance().recipeInterfaceImpl.linkAccounts({
+                    recipeUserId: recipeUserId,
+                    primaryUserId: sessionUser.id,
+                    userContext,
+                });
+
+                if (linkRes.status !== "OK") {
+                    throw new RecurseError();
+                }
+
+                let user = await getUser(recipeUserId.getAsString(), userContext);
+                if (user === undefined) {
+                    // linked user not found
+                    throw new SessionError({
+                        type: SessionError.UNAUTHORISED,
+                        message: "User not found",
+                    });
+                }
```

Comment:
remove this.. linkAccounts already returns a user object. Check all the places for mistakes like this.